handleinvite.go

```diff
@@ -0,0 +1,250 @@
+	}
+
+	// Check that the event is signed by the server sending the request.
+	redacted, err := verImpl.RedactEventJSON(input.InviteEvent.JSON())
+	if err != nil {
+		return nil, spec.BadJSON("The event JSON could not be redacted")
+	}
+
+	sender, err := spec.NewUserID(input.InviteEvent.Sender(), true)
+	if err != nil {
+		return nil, spec.BadJSON("The event JSON contains an invalid sender")
+	}
+	verifyRequests := []VerifyJSONRequest{{
+		ServerName:             sender.Domain(),
+		Message:                redacted,
+		AtTS:                   input.InviteEvent.OriginServerTS(),
+		StrictValidityChecking: true,
+	}}
+	verifyResults, err := input.Verifier.VerifyJSONs(input.Context, verifyRequests)
+	if err != nil {
+		util.GetLogger(input.Context).WithError(err).Error("keys.VerifyJSONs failed")
+		return nil, spec.InternalServerError{}
+	}
+	if verifyResults[0].Error != nil {
+		return nil, spec.Forbidden("The invite must be signed by the server it originated on")
+	}
+
+	// Sign the event so that other servers will know that we have received the invite.
+	signedEvent := input.InviteEvent.Sign(
+		string(input.InvitedUser.Domain()), input.KeyID, input.PrivateKey,
+	)
+
+	if signedEvent.StateKey() == nil {
+		util.GetLogger(input.Context).Error("invite must be a state event")
+		return nil, spec.InternalServerError{}
+	}
+
+	isKnownRoom, err := input.InviteQuerier.IsKnownRoom(input.Context, input.RoomID)
+	if err != nil {
+		util.GetLogger(input.Context).WithError(err).Error("failed querying known room")
+		return nil, spec.InternalServerError{}
+	}
+
+	inviteState := input.StrippedState
+	if len(inviteState) == 0 {
+		// "If they are set on the room, at least the state for m.room.avatar, m.room.canonical_alias, m.room.join_rules, and m.room.name SHOULD be included."
+		// https://matrix.org/docs/spec/client_server/r0.6.0#m-room-member
+		stateWanted := []StateKeyTuple{}
+		for _, t := range []string{
+			spec.MRoomName, spec.MRoomCanonicalAlias,
+			spec.MRoomJoinRules, spec.MRoomAvatar,
+			spec.MRoomEncryption, spec.MRoomCreate,
+		} {
+			stateWanted = append(stateWanted, StateKeyTuple{
+				EventType: t,
+				StateKey:  "",
+			})
+		}
+		if is, err := input.GenerateStrippedState(input.Context, input.RoomID, stateWanted, signedEvent); err == nil {
+			inviteState = is
+		} else {
+			util.GetLogger(input.Context).WithError(err).Error("failed querying known room")
+			return nil, spec.InternalServerError{}
+		}
+	}
+
+	logger := util.GetLogger(input.Context).WithFields(map[string]interface{}{
+		"inviter":  signedEvent.Sender(),
+		"invitee":  *signedEvent.StateKey(),
+		"room_id":  input.RoomID.String(),
+		"event_id": signedEvent.EventID(),
+	})
+	logger.WithFields(logrus.Fields{
+		"room_version":     signedEvent.Version(),
+		"room_info_exists": isKnownRoom,
+	}).Debug("processing incoming federation invite event")
+
+	if len(inviteState) == 0 {
```

Comment:
It shouldn't be possible if this is a locally known room (ie. there should be at least a create event). I can add a guard for that case since we shouldn't just be content with that.
It is possible if this isn't a known room since there would be no state for us to generate.